### ISSUE TITLE
Merge all the different things that call 'sendmail' to a common API

### ISSUE THREAD

**Issue by brong** [cyrusimap/cyrus-imapd#2182]
e.g. sieve notify, JMAP, calendar notifies.

**Comment by rsto** [comment]
I've spent some time looking at the current codebase (I'm writing this up here since we'll need it during development):

Currently, Cyrus supports two main flows how to send mail: either by executing sendmail in "normal" mode (other than -bs) or by using the smptclient API that I introduced for JMAP mail. The latter is a thin wrapper over either a TCP-based or sendmail-backed SMTP connection.

The configuration by the former approach is via the cyrus.imapd `sendmail` option, where as the latter is configured by `jmap_smtp_backend`. Ideally, we'd just have one option to configure how Cyrus sends mail and just one way to do so.

That requires us to make use of SMTP backed email sending across all places in Cyrus, be that CalDAV scheduling, Sieve, NNTP or JMAP. But the current smptclient API as used in JMAP is too low level for this!

Especially the detection and handling of SMTP extensions is crude (if not to say: broken. Note how e.g. the hardcoded check for the AUTH extension via the string `250-AUTH` [here](https://github.com/cyrusimap/cyrus-imapd/blob/master/imap/jmap_mail.c#L8377)).
We'll need better SMTP extension support for DSNs, at least.

Another data point: Note how currently the JMAP mail module takes care of properly dot-escaping mail messages ([here](https://github.com/cyrusimap/cyrus-imapd/blob/master/imap/jmap_mail.c#L8432)).

TL;DR: We shouldn't just refactor the current sendmail-backed mail sending modules to make use of SMTP, but rework the smtpclient API into a nicer, high-level API while doing so.

**Comment by rsto** [comment]
This randomly-assigned diceroll is highly relevant: https://github.com/cyrusimap/cyrus-imapd/issues/1637

**Comment by rsto** [comment]
This will land as https://github.com/cyrusimap/cyrus-imapd/pull/2225

**Comment by rsto** [comment]
Fixed with https://github.com/cyrusimap/cyrus-imapd/commit/e0d4a90c1787709578a04f4c236463bc41c82858 and https://github.com/cyrusimap/cassandane/commit/2519d4abbc618610576d016a98005586076f5ffc
